File: arith_evaluator/lib/parser.ml
open Ast
open Lexer

type st_elem = Texp of expression | Tunr of unary_op | Tbin of binary_op | Tlp

exception Parser_error of string

let uop_of_sym s =
  match s with
  | "-" -> UMINUS
  | "+" -> UPLUS
  | _ -> raise (Parser_error ("Invalid unary operator symbol: " ^ s))

let bop_of_sym s =
  match s with
  | "+" -> PLUS
  | "-" -> MINUS
  | "*" -> MULT
  | "/" -> DIV
  | "%" -> MOD
  | _ -> raise (Parser_error ("Invalid binary operator symbol: " ^ s))

let tsym s =
  try Tbin (bop_of_sym s) with Parser_error _ -> Tunr (uop_of_sym s)

let reduce pr = function
  | Texp e :: Tunr op :: st when priority_uop op >= pr ->
      Texp (ExpUnary (op, e)) :: st
  | Texp e1 :: Tbin op :: Texp e2 :: st when priority_bop op >= pr ->
      Texp (ExpBinary (e2, op, e1)) :: st
  | _ -> raise (Parser_error "Could not reduce stack")

let rec stack_or_reduce lexeme stack =
  match (lexeme, stack) with
  | Lint n, _ -> Texp (ExpInt n) :: stack
  | Lsym "(", _ -> Tlp :: stack
  | Lsym ")", Texp e :: Tlp :: st -> Texp e :: st
  | Lsym ")", _ -> stack_or_reduce lexeme (reduce 0 stack)
  | Lsym s, _ -> (
      let sym =
        match s with
        | "+" -> (
            match stack with Texp _ :: _ -> Tbin PLUS | _ -> Tunr UPLUS)
        | "-" -> (
            match stack with Texp _ :: _ -> Tbin MINUS | _ -> Tunr UMINUS)
        | _ -> tsym s
      in
      match sym with
      | Tunr op -> Tunr op :: stack
      | Tbin op -> (
          try stack_or_reduce lexeme (reduce (priority_bop op) stack)
          with Parser_error _ -> Tbin op :: stack)
      | _ -> raise (Parser_error "Invalid state"))
  | _, _ -> raise (Parser_error "Invalid state")

let rec reduce_all = function
  | [] -> raise (Parser_error "Invalid expression")
  | [ Texp ast ] -> ast
  | st -> reduce_all (reduce 0 st)

let parse_exp stop l =
  let p = ref 0 in
  let rec parse_one stack =
    let cl =
      p := l.pos;
      lex l
    in
    if not (stop cl) then parse_one (stack_or_reduce cl stack)
    else (
      l.pos <- !p;
      reduce_all stack)
  in
  parse_one []

File: arith_evaluator_rs/src/parser.rs
use super::ast::{BinaryOp, Expression, UnaryOp};
use std::error::Error;
use std::fmt;

#[derive(Debug)]
enum StackElem {
    Texp(Box<Expression>),
    Tunr(UnaryOp),
    Tbin(BinaryOp),
    Tlp,
}

#[derive(Debug)]
pub struct Parser;

impl Parser {
    pub fn new() -> Self {
        Parser {}
    }

    fn priority_uop(op: &UnaryOp) -> i32 {
        match op {
            UnaryOp::Uplus | UnaryOp::Uminus => 4,
        }
    }

    fn priority_bop(op: &BinaryOp) -> i32 {
        match op {
            BinaryOp::Mod => 1,
            BinaryOp::Plus | BinaryOp::Minus => 2,
            BinaryOp::Mult | BinaryOp::Div => 3,
        }
    }

    fn uop_of_string(sym: &str) -> Result<UnaryOp, ParserError> {
        match sym {
            "+" => Ok(UnaryOp::Uplus),
            "-" => Ok(UnaryOp::Uminus),
            _ => Err(ParserError::new(format!(
                "invalid symbol for unary operator: {}",
                sym
            ))),
        }
    }

    fn bop_of_string(sym: &str) -> Result<BinaryOp, ParserError> {
        match sym {
            "+" => Ok(BinaryOp::Plus),
            "-" => Ok(BinaryOp::Minus),
            "*" => Ok(BinaryOp::Mult),
            "/" => Ok(BinaryOp::Div),
            "%" => Ok(BinaryOp::Mod),
            _ => Err(ParserError::new(format!(
                "invalid symbol for binary operator: {}",
                sym
            ))),
        }
    }

    fn tsym(sym: &str) -> Result<StackElem, ParserError> {
        match Parser::bop_of_string(sym) {
            Ok(bop) => Ok(StackElem::Tbin(bop)),
            Err(_) => Ok(StackElem::Tunr(Parser::uop_of_string(sym)?)),
        }
    }

    fn reduce(&self, priority: usize, st: Vec<StackElem>) {
        match st.as_slice() {
            _ => todo!(),
        }

        todo!()
    }
}

#[derive(Debug)]
struct ParserError {
    message: String,
}

impl ParserError {
    pub fn new(message: String) -> Self {
        ParserError { message }
    }
}

impl Error for ParserError {}

impl fmt::Display for ParserError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.message)
    }
}
